### PR TITLE
Make ARSwitchboard return View Controllers

### PR DESCRIPTION
**Pull request by orta**
This lets us decouple the presentation of a VC from the the creation of the VC. Now all JLRouting is encapsulated in the ARSwitchboard

Removes the ARNavigationContainer, it did nothing. Removed the tests for it too.

I'm open to ideas on having a different class deal with the presentation logic, I don't really want every VC to know that it has to talk to a ARTopMenuVC. 

I'm also open to the idea that ARSwitchboard should be separated from the Eigen specific bits ( think the special casing for `/articles`, `/shows`, `/settings`, `/` ) which would help as we make VCs less aware of each other and Eigen on the whole.

Taking that even further, I'd be open to all of these:

```
/// Provide a simple API to load an ArtworkVC from a lot of different inputs
- (ARArtworkSetViewController *)loadArtworkSet:(NSArray *)artworkSet inFair:(Fair *)fair atIndex:(NSInteger)index;
- (ARArtworkSetViewController *)loadArtwork:(Artwork *)artwork inFair:(Fair *)fair;
- (ARArtworkSetViewController *)loadArtworkWithID:(NSString *)artworkID inFair:(Fair *)fair;

- (UIViewController *)loadBidUIForArtwork:(NSString *)artworkID inSale:(NSString *)saleID;

/// Load the auction results for an artwork on to the stack
- (ARAuctionArtworkResultsViewController *)loadAuctionResultsForArtwork:(Artwork *)artwork;
- (ARArtworkInfoViewController *)loadMoreInfoForArtwork:(Artwork *)artwork;

/// Load a Map VC
- (ARFairMapViewController *)loadMapInFair:(Fair *)fair;
- (ARFairMapViewController *)loadMapInFair:(Fair *)fair title:(NSString *)title selectedPartnerShows:(NSArray *)selectedPartnerShows;

- (ARArtistViewController *)loadArtistWithID:(NSString *)artistID;
- (UIViewController<ARFairAwareObject> *)loadArtistWithID:(NSString *)artistID inFair:(Fair *)fair;

/// Load a Partner Page in Martsy
- (UIViewController *)loadPartnerWithID:(NSString *)partnerID;

/// Load a Profile. Used to separate profiles with a fair from regular profiles.
- (UIViewController *)routeProfileWithID:(NSString *)profileID;

/// Load a Gene
- (ARGeneViewController *)loadGene:(Gene *)gene;
- (ARGeneViewController *)loadGeneWithID:(NSString *)geneID;

/// Load a fair booth
- (ARShowViewController *)loadShow:(PartnerShow *)show fair:(Fair *)fair;
- (ARShowViewController *)loadShow:(PartnerShow *)show;
- (ARShowViewController *)loadShowWithID:(NSString *)showID;
- (ARShowViewController *)loadShowWithID:(NSString *)showID fair:(Fair *)fair;

```

Being a class category specific to Eigen. Then if/when we do pod-based VCs we can expose ARSwitchboard for others to hook into.

### CODE REVIEW

File: Artsy/App/ARSwitchBoard.m [hunk 1]
```diff
@@ -75,140 +115,112 @@ - (id)init
...
-        [[ARTopMenuViewController sharedController] pushViewController:viewController];
-
-        return YES;
+        return [sself loadFairGuideWithFair:fair];
     }];
 
-    [self.routes addRoute:@"/:profile_id/browse/artist/:id" handler:^BOOL(NSDictionary *parameters) {
-
+    [self.routes addRoute:@"/:profile_id/browse/artist/:id" handler:JLRouteParams {
         if ([UIDevice isPad]) { return NO; }
```
Comment (alloy) [inline Artsy/App/ARSwitchBoard.m]: Incorrect return type.

**Comment by alloy** [comment]
>I'm open to ideas on having a different class deal with the presentation logic, I don't really want every VC to know that it has to talk to a ARTopMenuVC.
>
> I'm also open to the idea that ARSwitchboard should be separated from the Eigen specific bits ( think the special casing for /articles, /shows, /settings, / ) which would help as we make VCs less aware of each other and Eigen on the whole.

I agree that the routed VCs should not need to know about external concerns such as the top menu VC, but I think it’s fine for the switchboard to know about it, as long as it’s a class that lives in Eigen. We don’t know for sure yet if this exact same switchboard is usable outside of Eigen, so no need to abstract too much imo.

File: Artsy/App/ARSwitchBoard.m (hunk 1 above)
Comment (orta) [inline Artsy/App/ARSwitchBoard.m]: better have to make this not run on iPad instead of the other way anyway, good call

**Comment by orta** [comment]
Agree, taking it pragmatic!

File: Artsy/App/ARSwitchBoard.m [hunk 2]
```diff
@@ -353,29 +171,25 @@ - (UIViewController *)loadURL:(NSURL *)url fair:(Fair *)fair
 {
     NSParameterAssert(url);
 
-    // May be nil by the end of the method
-    UIViewController *viewController;
-
+    /// Is it an Artsy URL, or a purely relative path?
     if ([ARRouter isInternalURL:url] || url.scheme == nil) {
+        /// Normalize URL ( e.g. m.artsy.net -> stanging-m.artsyn.net
```
Comment (orta) [inline Artsy/App/ARSwitchBoard.m]: typo

File: Artsy/View_Controllers/App_Navigation/ARTopMenuViewController.m [hunk 3]
```diff
@@ -139,6 +132,27 @@ - (void)viewDidLoad
...
+        @(ARTopTabControllerIndexShows) : @"/shows",
+        @(ARTopTabControllerIndexNotifications) : @"/works-for-you",
+    };
+
+    for (NSNumber *tabIndex in menuToPaths.keyEnumerator) {
+        [switchboard registerPathCallbackAtPath:menuToPaths[tabIndex] callback:^id _Nullable(NSDictionary *_Nullable parameters) {
+            [self presentRootViewControllerAtIndex:tabIndex.integerValue animated:NO];
+            return self.rootNavigationController.topViewController;
+        }];
+    }
```
Comment (orta) [inline Artsy/View_Controllers/App_Navigation/ARTopMenuViewController.m]: Could UIKeyCommands also come in here?

File: Artsy_Tests/App_Tests/ARSwitchBoardTests.m [hunk 4]
```diff
@@ -154,115 +155,95 @@ @interface ARTopMenuNavigationDataSource (Tests)
...
+//                id viewController = [switchboard routeInternalURL:[NSURL URLWithString:@"http://artsy.net/shows"] fair:nil];
+//                expect(viewController).to.equal(navigationController.rootViewController);
+//            });
+//
+//            it(@"routes articles", ^{
+//                [[[controllerMock expect] andReturn:navigationController] rootNavigationControllerAtIndex:ARTopTabControllerIndexMagazine];
+//
+//                id viewController = [switchboard routeInternalURL:[NSURL URLWithString:@"http://artsy.net/articles"] fair:nil];
+//                expect(viewController).to.equal(navigationController.rootViewController);
+//            });
```
Comment (orta) [inline Artsy_Tests/App_Tests/ARSwitchBoardTests.m]: these should go to ARTopVC

**Comment by orta** [comment]
Alright, this is :+1: to go now

Ended up with the ARTopViewController registering paths for it's menu items, felt hacky keeping track of indexes etc in the Switchboard. Plus this was a pattern I wanted to look at for Auctions, and it looks like it works.

File: Artsy/App/ARSwitchBoard.m [hunk 5]
```diff
@@ -62,273 +49,106 @@ - (id)init
...
+    if (![UIDevice isPad]) {
+        [self.routes addRoute:@"/:profile_id/for-you" handler:JLRouteParams {
+            __strong typeof (wself) sself = wself;
+            Fair *fair = [parameters[@"fair"] isKindOfClass:Fair.class] ? parameters[@"fair"] : nil;
+            return [sself loadFairGuideWithFair:fair];
+        }];
+    }
 
+    [self.routes addRoute:@"/:profile_id/browse/artist/:id" handler:JLRouteParams {
         if ([UIDevice isPad]) { return NO; }
```
Comment (alloy) [inline Artsy/App/ARSwitchBoard.m]: Still the wrong return type here. Isn’t Xcode telling you about this?

File: Artsy/App/ARSwitchBoard.m [hunk 6]
```diff
@@ -62,273 +49,106 @@ - (id)init
...
-        [[ARTopMenuViewController sharedController] pushViewController:viewController];
-        return YES;
-    }];
-
-    [self.routes addRoute:@"/categories" handler:^BOOL(NSDictionary *parameters) {
-        ARBrowseCategoriesViewController *viewController = [[ARBrowseCategoriesViewController alloc] init];
-        [[ARTopMenuViewController sharedController] pushViewController:viewController];
-        return YES;
+    [self.routes addRoute:@"/categories" handler:JLRouteParams {
+        return [[ARBrowseCategoriesViewController alloc] init];;
```
Comment (alloy) [inline Artsy/App/ARSwitchBoard.m]: Needs more semi-colons.

File: Artsy/App/ARSwitchBoard.m (hunk 5 above)
Comment (orta) [inline Artsy/App/ARSwitchBoard.m]: nope, interesting, wonder if it just coalesces into a nil, will edit

File: Artsy/App/ARSwitchBoard.m (hunk 6 above)
Comment (orta) [inline Artsy/App/ARSwitchBoard.m]: Giphy

**Comment by alloy** [comment]
Alright, awesome :+1:

Last thing is some changelog love :)

**Comment by orta** [comment]
alright, changelog shipped

**Comment by orta** [comment]
Merging, and assuming that a CHANGELOG entry isn't gonna break a build

**Pull request merged**